File: Python/gol.py
import time

def save_pbm(univ, w, h, iter_count):
    filename = f"gol_{iter_count}.pbm"
    try:
        with open(filename, "w") as f:
            f.write(f"P1\n{w} {h}\n")
            for row in univ:
                # Replica o formato exato do C: "%d "
                f.write("".join(f"{cell} " for cell in row) + "\n")
    except IOError as e:
        print(f"Erro ao criar o arquivo PBM: {e}")

def evolve(univ, w, h):
    # Cria uma matriz temporária para armazenar o próximo estado
    new = [[0 for _ in range(w)] for _ in range(h)]

    # Varredura completa da matriz
    for y in range(h):
        for x in range(w):
            n = 0
            
            # Laços internos para inspecionar a vizinhança 3x3
            for y1 in range(y - 1, y + 2):
                for x1 in range(x - 1, x + 2):
                    # Verifica se a célula vizinha está viva, usando módulo para tabuleiro infinito
                    if univ[(y1 + h) % h][(x1 + w) % w]:
                        n += 1

            # Desconta a própria célula caso ela esteja viva
            if univ[y][x]:
                n -= 1
            
            # Aplica as regras de sobrevivência e nascimento
            if n == 3 or (n == 2 and univ[y][x]):
                new[y][x] = 1
            else:
                new[y][x] = 0

    # Segunda passagem para copiar o estado calculado de volta para a matriz original
    for y in range(h):
        for x in range(w):
            univ[y][x] = new[y][x]

def game(w, h, max_iter):
    # Inicializa o tabuleiro: Uma cruz perfeita cruzando a matriz ao meio
    # Usa w // 2 e h // 2 para encontrar a linha e coluna centrais
    univ = [[1 if (x == w // 2 or y == h // 2) else 0 for x in range(w)] for y in range(h)]

    save_interval = 500
    # Laço iterativo simulando o jogo
    for iter_count in range(max_iter + 1):
        if iter_count % save_interval == 0:
            save_pbm(univ, w, h, iter_count)
        evolve(univ, w, h)

def main():
    # Dimensões e iterações fixas
    w = 500
    h = 500
    max_iter = 5000
    
    # Captura o tempo EXATAMENTE ANTES do processamento iniciar
    start = time.perf_counter()

    # Chama a função principal
    game(w, h, max_iter)

    # Captura o tempo EXATAMENTE APÓS o processamento terminar
    end = time.perf_counter()

    # Calcula o tempo total
    time_taken = end - start

    # Imprime o tempo de validação
    print(f"Tempo interno de execucao: {time_taken:f} segundos")

if __name__ == "__main__":
    main()

File: C/gol.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

// Função para gerar o arquivo PBM
void save_pbm(void *u, int w, int h, int iter)
{
   int (*univ)[w] = u;
   char filename[64];
   
   // Formata o nome do arquivo para gol_nn.pbm (ex: gol_0, gol_500, gol_1000)
   sprintf(filename, "gol_%d.pbm", iter);
   
   FILE *f = fopen(filename, "w");
   if (!f) {
      perror("Erro ao criar o arquivo PBM");
      return;
   }

   // Cabeçalho e dados PBM no formato P1 (ASCII monocromático)
   fprintf(f, "P1\n%d %d\n", w, h);
   for (int y = 0; y < h; y++) {
      for (int x = 0; x < w; x++) {
         fprintf(f, "%d ", univ[y][x]);
      }
      fprintf(f, "\n");
   }
   fclose(f);
}

// Função principal que calcula a próxima geração do Game of Life.
void evolve(void *u, int w, int h)
{
   unsigned (*univ)[w] = u;
   unsigned new[h][w];

   for (int y = 0; y < h; y++) {
      for (int x = 0; x < w; x++) {
         int n = 0;
         
         for (int y1 = y - 1; y1 <= y + 1; y1++) {
            for (int x1 = x - 1; x1 <= x + 1; x1++) {
               if (univ[(y1 + h) % h][(x1 + w) % w]) {
                  n++;
               }
            }
         }

         if (univ[y][x]) n--;
         
         new[y][x] = (n == 3 || (n == 2 && univ[y][x]));
      }
   }

   for (int y = 0; y < h; y++) {
      for (int x = 0; x < w; x++) {
         univ[y][x] = new[y][x];
      }
   }
}

// A função game recebe largura, altura da matriz e o total de iterações 
void game(int w, int h, int max_iter)
{
   unsigned univ[h][w];
   
   // Inicializa o tabuleiro: Uma cruz perfeita cruzando a matriz ao meio
   for (int y = 0; y < h; y++) {
      for (int x = 0; x < w; x++) {
         // Ativa a célula se ela estiver na linha do meio OU na coluna do meio
         if (x == w / 2 || y == h / 2) {
            univ[y][x] = 1;
         } else {
            univ[y][x] = 0;
         }
      }
   }
   
   const int save_interval = 500; // Salva a cada 500 iterações
   for (int iter = 0; iter <= max_iter; iter++) {
      if (iter % save_interval == 0) {
         save_pbm(univ, w, h, iter);
      }
      evolve(univ, w, h);
   }
}

int main(void)
{
   int w = 500;
   int h = 500;
   int max_iter = 5000;

   struct timespec start, end;

   // Captura o tempo EXATAMENTE ANTES do processamento iniciar
   clock_gettime(CLOCK_MONOTONIC, &start);

   game(w, h, max_iter);

   // Captura o tempo EXATAMENTE APÓS o processamento terminar
   clock_gettime(CLOCK_MONOTONIC, &end);

   double time_taken = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1000000000.0;

   printf("Tempo interno de execucao: %f segundos\n", time_taken);
   
   return 0;
}
